### PR TITLE
feat: GLM-5.2-NVFP4 support to AIC (collector + modeling)

### PR DESCRIPTION
**Pull request by YijiaZhao**
#### Overview:

  Add GLM-5.2-NVFP4 support to AIC — both perf-data collection (collector) and
  modeling (SDK) — on the existing `GlmMoeDsaForCausalLM` / DEEPSEEKV32 path.
  GLM-5 and GLM-5.2 NVFP4 both run `cli default` end-to-end on
  `b200_sxm` / `sglang` / `0.5.13`. The DSA attention module is modeled in
  **bf16** (NVFP4 quantizes only MoE expert linears; `self_attn` is excluded),
  and GLM-5.2's shared-index "skip-indexer" layers are amortized using directly
  collected reuse-layer data.

  #### Details:

  **Modeling (`src/aiconfigurator/sdk`)**
  - DSA attention gemm is bf16, not nvfp4: `self_attn` (incl. indexer) is in
    `hf_quant_config.exclude_modules`, so `deepseek_v32` forces
    `dsa_gemm_quant_mode=bfloat16`.
  - Skip-indexer amortization: GLM-5.2 shares one DSA topk index across
    `index_topk_freq` layers (1 full + rest reuse). Per-layer cost is
    `full_frac*full + (1-full_frac)*skip`, where `full_frac` is the EXACT
    indexer-computing fraction replicating sglang `dsa_layer_skips_topk`
    (GLM-5.2 = 21/78 = 0.2692, not `1/freq`=0.25). `full_frac==1.0`
    (GLM-5 / DeepSeek-V3.2) → pure full, behavior unchanged.
  - Full and skip rows share ONE perf file, split by the `op_name` column via
    `op_kind="full"/"skip"`; the loader is model-agnostic (GLM-5.2 data serves
    GLM-5). `clear_cache` now also clears the skip caches (ctx + gen).

  **Collector (`collector/sglang`)**
  - New `dsa_context/generation_module_skip_indexer` ops: patch out the per-layer
    indexer (`skip_topk` + reuse a captured `topk_indices`) to time the
    reuse-layer cost; rows land in the main perf file tagged `_skip_indexer`.
    The skip flag is threaded as an explicit subprocess arg (mirroring
    `is_prefill`) — no env var.
  - `dev-glm52-nvfp4` image compatibility, all backward-compatible with 0.5.13:
    KV pool + attn-backend init, `set_extend_range` guards, piecewise-config
    getattr guards, module-piecewise replay fallback-to-eager, `layer_types`
    strip. `dsa_generation` smoke passes on **both** 0.5.13 and dev images.
  - Same-precision DSA dedup: when several same-`(arch, quant)` checkpoints are
    requested, collect once on the longest-context representative (GLM-5.2) and
    add the shorter siblings' `max_position` ceiling points (ctx + gen) so their
    near-max queries hit data instead of extrapolating.
#### Overview:

  Add GLM-5.2-NVFP4 support to AIC — both perf-data collection (collector) and
  modeling (SDK) — on the existing `GlmMoeDsaForCausalLM` / DEEPSEEKV32 path.
  GLM-5 and GLM-5.2 NVFP4 both run `cli default` end-to-end on
  `b200_sxm` / `sglang` / `0.5.13`. The DSA attention module is modeled in
  **bf16** (NVFP4 quantizes only MoE expert linears; `self_attn` is excluded),
  and GLM-5.2's shared-index "skip-indexer" layers are amortized using directly
  collected reuse-layer data.

  #### Details:

  **Modeling (`src/aiconfigurator/sdk`)**
  - DSA attention gemm is bf16, not nvfp4: `self_attn` (incl. indexer) is in
    `hf_quant_config.exclude_modules`, so `deepseek_v32` forces
    `dsa_gemm_quant_mode=bfloat16`.
  - Skip-indexer amortization: GLM-5.2 shares one DSA topk index across
    `index_topk_freq` layers (1 full + rest reuse). Per-layer cost is
    `full_frac*full + (1-full_frac)*skip`, where `full_frac` is the EXACT
    indexer-computing fraction replicating sglang `dsa_layer_skips_topk`
    (GLM-5.2 = 21/78 = 0.2692, not `1/freq`=0.25). `full_frac==1.0`
    (GLM-5 / DeepSeek-V3.2) → pure full, behavior unchanged.
  - Full and skip rows share ONE perf file, split by the `op_name` column via
    `op_kind="full"/"skip"`; the loader is model-agnostic (GLM-5.2 data serves
    GLM-5). `clear_cache` now also clears the skip caches (ctx + gen).

  **Collector (`collector/sglang`)**
  - New `dsa_context/generation_module_skip_indexer` ops: patch out the per-layer
    indexer (`skip_topk` + reuse a captured `topk_indices`) to time the
    reuse-layer cost; rows land in the main perf file tagged `_skip_indexer`.
    The skip flag is threaded as an explicit subprocess arg (mirroring
    `is_prefill`) — no env var.
  - `dev-glm52-nvfp4` image compatibility, all backward-compatible with 0.5.13:
    KV pool + attn-backend init, `set_extend_range` guards, piecewise-config
    getattr guards, module-piecewise replay fallback-to-eager, `layer_types`
    strip. `dsa_generation` smoke passes on **both** 0.5.13 and dev images.
  - Same-precision DSA dedup: when several same-`(arch, quant)` checkpoints are
    requested, collect once on the longest-context representative (GLM-5.2) and
    add the shorter siblings' `max_position` ceiling points (ctx + gen) so their
    near-max queries hit data instead of extrapolating.

## Summary by CodeRabbit

* **New Features**
  * Added support for the NVIDIA **GLM-5.2 NVFP4** model variant (including updated model/config presets).
  * Added **DSA skip-indexer (reuse-layer)** benchmark variants for both context and generation.
* **Bug Fixes**
  * Improved routing and amortized blending between full vs skip-indexer performance data.
  * Enhanced runtime robustness across sglang/version differences and safer parameter handling.
* **Refactor**
  * Centralized full/skip data loading and query amortization behavior for DSA modules.
* **Tests**
  * Updated and added unit tests covering GLM NVFP4 selection/deduping and skip-indexer table routing.

### CODE REVIEW

**Review by tianhaox — COMMENTED** [review]
Thanks for adding GLM-5.2 NVFP4 support. I reviewed the collector and Python SDK paths, and I think the following correctness gaps should be addressed before merge:

1. **Targeted sparse collection selects the wrong checkpoint.** In `_selected_glm5_models()`, a run targeted only at `nvidia/GLM-5.2-NVFP4` falls through to `GLM5_DEFAULT_MODEL` and collects `nvidia/GLM-5-NVFP4`. This loses the GLM-5.2 1M-context coverage. A single selected model should be returned directly; only the multi-model case should canonicalize to the longest-context representative.

2. **The context skip-indexer pass can record full-indexer latency.** On SGLang versions where the legacy piecewise replay path is available, `call_target` uses `model_runner.forward()` and bypasses `_skip_kwargs()`. The row is still tagged `dsa_context_module_skip_indexer`, so full-layer timings can be written into the skip table. Please force the skip pass to eager/module execution, or explicitly propagate `skip_topk` and `prev_topk_indices` through replay.

3. **Shared experts are modeled with the wrong quantization.** The GLM-5.2 quant config excludes `mlp.shared_experts*` from NVFP4, but the SDK still creates the shared-expert GEMMs with the global `gemm_quant_mode=nvfp4`. Those GEMMs should use BF16 (or a dedicated shared-expert quant mode), otherwise the model systematically underestimates their latency.

4. **`skip_indexer` does not change the analytic SOL.** It only selects a different table; the context and generation `get_sol()` implementations still include indexer projections, logits, and index-cache traffic. This makes full and skip SOL identical, and also gives incorrect EMPIRICAL/HYBRID calibration and boundary extrapolation. Please add skip-aware SOL accounting.

5. **The producer/consumer contract is not backend-safe.** `dsa_full_layer_fraction=21/78` is applied for every backend, while skip-indexer producers exist only for SGLang. vLLM and TensorRT-LLM can therefore pass full-table validation and fail later when the SDK requests a skip table. Please either restrict this amortization/support to a backend with skip data, or add equivalent producers and coverage for the other backends. The same validation should require both full and skip data for GLM-5.2.

6. **CP and MTP composition need small corrections.** A CP skip layer should not pay the indexer-key all-gather, because the indexer is not executed. For generation, the scale includes the extra MTP layer, but the fraction remains `21/78`; since the MTP layer uses `skip_topk=True`, the generation fraction should reflect 21 full layers across 79 executed layers.

7. **Injected ceiling shapes bypass the context environment filter.** The base shapes are filtered first, then ceiling points are appended unconditionally. A single-shape targeted collection can therefore expand back into many cases. Please apply the same filter after merging the ceiling shapes.

The focused unit tests, Ruff checks, and existing CI are green, but there are currently no direct regression tests for the actual full/skip query routing, `op_kind` row separation, full-layer fraction, backend data availability, or mixed result provenance. Adding those tests would make the intended behavior much safer to maintain.

File: src/aiconfigurator/sdk/operations/dsa.py [hunk 1]
```diff
@@ -1101,28 +1199,44 @@ def load_data(cls, database: PerfDatabase) -> None:
...
+                )
+                cls._extrapolate(cls._skip_data_cache[key])
+            else:
+                cls._skip_data_cache[key] = None
+
         if "_generation_dsa_module_data" not in database.__dict__:
             database._generation_dsa_module_data = cls._data_cache[key]
             database._raw_generation_dsa_module_data = cls._raw_data_cache[key]
+        if "_generation_dsa_module_skip_data" not in database.__dict__:
+            database._generation_dsa_module_skip_data = cls._skip_data_cache[key]
```
[the remark this answers is not in the document]
Comment (YijiaZhao) [reply src/aiconfigurator/sdk/operations/dsa.py:1257]: Fixed. `query_generation_dsa_module` now reads a dedicated `_raw_generation_dsa_module_skip_data`, populated alongside the full raw cache in `load_data`, so skip-indexer boundary extrapolation anchors to the skip rows rather than the full ones.

File: collector/sglang/glm5_dsa_sparse_modules.py [hunk 2]
```diff
@@ -59,6 +59,21 @@
...
+    try:
+        try:
+            from collector.sglang.collect_mla_module import get_mla_module_model_specs
+        except ModuleNotFoundError:
+            from collect_mla_module import get_mla_module_model_specs
+        paths = {s.model_path for s in get_mla_module_model_specs(attention_type="dsa")}
+        if {"nvidia/GLM-5-NVFP4", "nvidia/GLM-5.2-NVFP4"} <= paths:
+            return ["nvidia/GLM-5.2-NVFP4"]
+    except Exception:
+        pass
```
[the remark this answers is not in the document]
Comment (YijiaZhao) [reply collector/sglang/glm5_dsa_sparse_modules.py]: Fixed in 89f83e66. Removed the blanket `except Exception: pass` so discovery errors propagate instead of silently degrading to `GLM5_DEFAULT_MODEL`. `_selected_glm5_models()` now returns `GLM-5.2-NVFP4` whenever it's present in the configured DSA specs (`apply_model_filter=False`, so a `COLLECTOR_MODEL_PATH` pin to GLM-5.2 still resolves to it), and only the genuine absence of any GLM-5.x spec falls back to the default.

File: collector/sglang/glm5_dsa_sparse_modules.py [hunk 3]
```diff
@@ -420,13 +437,29 @@ def _env(cases):
...
+    # near-max queries interpolate within data instead of extrapolating across
+    # the coarse prefix grid.
+    have = set(shapes)
+    for cover in _dsa_ceiling_max_positions(model_path):
+        for isl in sweep.context_sequence_lengths:
+            ceil_prefix = cover - isl
+            for bs in sweep.context_batch_sizes:
+                key = (ceil_prefix, isl, bs)
+                if ceil_prefix > 0 and key not in have and _valid(bs, isl, ceil_prefix):
+                    shapes.append(key)
```
[the remark this answers is not in the document]
Comment (YijiaZhao) [reply collector/sglang/glm5_dsa_sparse_modules.py]: Fixed in 89f83e66. The injected ceiling `(bs, isl, prefix)` candidates are now collected and passed through the same `_env` (`AIC_DSA_CONTEXT_*`) filter as the base shapes before being appended, so a targeted collector run can no longer pick up extra ceiling shapes.

File: tests/unit/sdk/database/conftest.py [hunk 4]
```diff
@@ -157,7 +157,9 @@ def _patch_all_loaders_and_yaml(monkeypatch) -> None:
 
     for name, (target_module, default_value) in _LOADER_STUBS.items():
         ret = overrides.get(name, default_value)
-        monkeypatch.setattr(f"{target_module}.{name}", lambda path, _r=ret: _r)
+        # Accept arbitrary extra args/kwargs so loaders with optional params
+        # (e.g. load_*_dsa_module_data(path, op_kind="full")) stub cleanly.
+        monkeypatch.setattr(f"{target_module}.{name}", lambda path, *args, _r=ret, **kwargs: _r)
```
[the remark this answers is not in the document]
Comment (YijiaZhao) [reply tests/unit/sdk/database/conftest.py]: Fixed in 89f83e66. The DSA module loader stubs now honor `op_kind`: when an override supplies op_kind-keyed data (`{"full": ..., "skip": ...}`) the stub returns the matching table (the default `None` is unchanged). Added `test_skip_indexer_routes_to_skip_table`, which sets distinct full/skip tables (10.0 vs 4.0) and asserts `skip_indexer=True` reads the skip table — a wrong wiring now fails instead of silently passing.

**Comment by YijiaZhao** [comment]
@tianhaox thanks for the thorough review — all addressed in 89f83e66 (plus the earlier gen-skip raw-cache fix). Point by point:

**1. Targeted sparse collection selects the wrong checkpoint.** Fixed. `_selected_glm5_models()` now resolves to `GLM-5.2-NVFP4` whenever it's in the configured DSA specs (via `apply_model_filter=False`, independent of a `COLLECTOR_MODEL_PATH` pin), so a run targeted only at GLM-5.2 no longer falls through to the default. The silent `except` was removed.

**2. Context skip-indexer pass can record full-indexer latency.** Fixed. When `skip_indexer` is active in `_run_prefill`, we now force `use_full_model_piecewise_replay = use_module_piecewise_replay = False`. The replay path calls `model_runner.forward()` and can't thread `_skip_kwargs()`, so only the eager `call_attention_module` path (which sets `skip_topk` + reuses `prev_topk_indices`) runs — the skip rows now genuinely exclude the indexer. Generation was already eager + `_skip_kwargs()`.

**3. Shared experts modeled with the wrong quantization.** Confirmed against the real checkpoint and fixed: `layers.10.mlp.shared_experts.gate_proj.weight` is **BF16** ([2048, 6144], no `weight_scale`/`weight_scale_2`/`input_scale`), while routed `experts.0.gate_proj.weight` is **U8** NVFP4 with the full scale set. Added `_shared_experts_excluded_from_quant()`; when the checkpoint excludes `mlp.shared_experts*`, the shared-expert GEMMs now build at `bf16` (via `dsa_shared_expert_quant_mode`) instead of the global `nvfp4`. Routed experts stay NVFP4.

**4. `skip_indexer` does not change the analytic SOL.** Fixed for context: `get_sol` now closes over `skip_indexer` and a skip layer zeroes `indexer_logits_ops` and `indexer_cache_bytes` (it reuses a sibling's topk and never runs the per-layer indexer). Generation already excludes the mqa/topk deltas for skip layers.

**5. Producer/consumer contract not backend-safe.** Fixed. `dsa_full_layer_fraction` is now forced to `1.0` (all-full) for any backend other than sglang, since the skip-indexer producer rows exist only for the sglang collector. The model still carries `index_topk_freq`; we just can't model the skip saving without data, so non-sglang backends never query a nonexistent skip table.

**6. CP and MTP composition.** CP part fixed: a CP skip layer no longer pays the indexer-key all-gather (`ag_kv` is zeroed when `skip_indexer`), keeping only the MLA compressed-KV/LSE gather. For the **MTP** part — could you point to the specific composition you'd like corrected? I want to match your intent rather than guess.

**7. Injected ceiling shapes bypass the context env filter.** Fixed — ceiling points now go through the same `_env` filter as the base shapes.

File: collector/sglang/collect_mla_module.py [hunk 5]
```diff
@@ -1657,6 +1871,46 @@ def _capture_one_batch_size_with_module_batch(runner, num_tokens: int):
...
+        def _skip_kwargs():
+            # Once a real topk index is captured, bypass the indexer for all
+            # subsequent (timed) forwards by forcing skip_topk + reusing it.
+            if _skip_indexer and _skip_state["prev_topk"] is not None:
+                attention_module.skip_topk = True
+                if _skip_state["hook"] is not None:
+                    _skip_state["hook"].remove()
+                    _skip_state["hook"] = None
+                return {"prev_topk_indices": _skip_state["prev_topk"]}
+            return {}
```
[the remark this answers is not in the document]
Comment (YijiaZhao) [reply collector/sglang/collect_mla_module.py:1917]: Fixed in e95a7260. The skip-indexer pass now aborts (raises) before timing/logging if the attention module has no `indexer`, and again after warmup if no topk was captured (`_skip_state["prev_topk"] is None`) — in both prefill and decode. A skip row is only logged when a real topk capture succeeded.

File: collector/sglang/collect_mla_module.py [hunk 6]
```diff
@@ -2161,16 +2462,44 @@ def run_mla_module(
...
+        _sweep = get_mla_module_sweep_spec("sglang")
+        _max_pos = _model_max_position_embeddings(model_path)
+        cases = [(bs, seq_len, ip, 0) for bs, seq_len, ip in base_cases]
+        _have = {(bs, sl) for bs, sl, _ in base_cases}
+        _extra_kv = set(_sweep.context_prefix_lengths)
+        for _cover_pos in _DSA_CEILING_MAX_POSITIONS:
+            _extra_kv.add(_cover_pos - 1)
+        for _sl in sorted(_extra_kv):
+            if _sl <= 0 or (_max_pos is not None and _sl > _max_pos - 1):
+                continue
```
[the remark this answers is not in the document]
Comment (YijiaZhao) [reply collector/sglang/collect_mla_module.py:2525]: Fixed in e95a7260. The decode extra-KV ceiling expansion is now gated on `_max_pos is not None`, mirroring the context path, so an unknown model context limit no longer injects the ~1M-KV DSA ceilings.

**Review by tianhaox — APPROVED** [review]
